File: ElementPathChildrenCollection.cs

/*****************************************************************************
 *                                                                           *
 *  ElementPathChildrenCollection.cs                                         *
 *  25 November 2006                                                         *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries, Inc.                            *
 *                                                                           *
 *  Converted from C++ to C# on 10 November 2007                             *
 *                                                                           *
 *  An unordered collection of Path objects.                                 *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Collections;

namespace MetaphysicsIndustries.Crystalline
{
    [Serializable]
    public abstract class ElementPathChildrenCollection : ICollection<Path>//, IDisposable
	{
        protected ElementPathChildrenCollection() { }

		public ElementPathChildrenCollection(Element element)
		{
			_parent = element;
			_set = new Set<Path>();
		}

        //public virtual void Dispose()
        //{
        //    this.Clear();
        //    //_set.Dispose();
        //}

		public virtual void Add(Path path)
		{
			if (!this.Contains(path))
			{
				
				this.SetParentElementOfPath(path, null);
				_set.Add(path);
				
				this.SetParentElementOfPath(path, _parent);
			}
		}

		public virtual void Clear()
		{
			List<Path>	list;
			list = new List<Path>();
			foreach (Path path in this)
			{
				list.Add(path);
			}
			foreach (Path path in list)
			{
				this.Remove(path);
			}
			list.Clear();
			_set.Clear();
		}

		public virtual bool Contains(Path path)
		{
			return _set.Contains(path);
		}

		public virtual bool Remove(Path path)
		{
			bool	b;
			b = _set.Remove(path);
			if (b && this.GetParentElementOfPath(path) == _parent)
			{
				
				this.SetParentElementOfPath(path, null);
			}
			return b;
		}

		public virtual void CopyTo(Path[] r, int i)
		{
			_set.CopyTo(r, i);
		}

		public virtual IEnumerator<Path> GetEnumerator()
		{
			return _set.GetEnumerator();
		}

		public virtual bool IsReadOnly
		{
			get
			{
				return _set.IsReadOnly;
			}
		}

		public virtual int Count
		{
			get
			{
				return _set.Count;
			}
		}

        protected abstract void SetParentElementOfPath(Path path, Element element);

		protected abstract Element GetParentElementOfPath(Path path);

        private Element _parent;
        private Set<Path> _set;

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}

File: Crystalline/ElementPathChildrenCollection.cpp

//////////// source file ////////////


/*****************************************************************************
 *                                                                           *
 *  ElementPathChildrenCollection.cpp                                        *
 *  25 November 2006                                                         *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries                                  *
 *                                                                           *
 *  An unordered collection of Path objects.                                 *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "ElementPathChildrenCollection.h"
#include "Element.h"
#include "Path.h"



NAMESPACE_START;


ElementPathChildrenCollection::ElementPathChildrenCollection(Element^ element)
{
	this->parent = element;
	this->set = gcnew Set<Path^>;
}

ElementPathChildrenCollection::~ElementPathChildrenCollection(void)
{
	this->Clear();
	delete this->set;
}

void ElementPathChildrenCollection::Add(Path^ path)
{
	if (!this->Contains(path))
	{
		//path->ParentElement = nullptr;
		this->SetChildsParent(path, nullptr);
		this->set->Add(path);
		//path->ParentElement = this->parent;
		this->SetChildsParent(path, this->parent);
	}
}

bool ElementPathChildrenCollection::Contains(Path^ path)
{
	return this->set->Contains(path);
}

bool ElementPathChildrenCollection::Remove(Path^ path)
{
	bool	b;

	b = this->set->Remove(path);
//	if (b && path->ParentElement == this->parent)
	if (b && this->GetChildsParent(path) == this->parent)
	{
		//path->ParentElement = nullptr;
		this->SetChildsParent(path, nullptr);
	}

	return b;
}

void ElementPathChildrenCollection::Clear(void)
{
	List<Path^>^	list;

	list = gcnew List<Path^>;
	for each (Path^ path in this)
	{
		list->Add(path);
	}

	for each (Path^ path in list)
	{
		this->Remove(path);
	}

	list->Clear();
	this->set->Clear();
}

void ElementPathChildrenCollection::CopyTo(array<Path^>^ r, int i)
{
	this->set->CopyTo(r, i);
}

IEnumerator<Path^>^ ElementPathChildrenCollection::GetEnumerator(void)
{
	return this->set->GetEnumerator();
}

int ElementPathChildrenCollection::Count::get(void)
{
	return this->set->Count;
}

bool ElementPathChildrenCollection::IsReadOnly::get(void)
{
	return this->set->IsReadOnly;
}

System::Collections::IEnumerator^ ElementPathChildrenCollection::GetEnumerator2(void)
{
	return this->GetEnumerator();
}



NAMESPACE_END;



